auth-source-pass.el

```diff
@@ -265,33 +235,105 @@ inside USER by giving priority to USER.  Same for PORT."
+   (auth-source-pass--applicable-entries hostname user port)
+   :initial-value nil))
+
+(defun auth-source-pass--select-from-entry-set (entry-set user)
+  "Return best matching password-store entry data from ENTRY-SET.
+
+If USER is non nil, give precedence to entries containing a user field
+matching USER."
+  (cl-reduce
+   (lambda (result entry)
+     (let ((entry-data (auth-source-pass-parse-entry entry)))
+       (cond ((equal (auth-source-pass--get-attr "user" result) user)
+              result)
+             ((equal (auth-source-pass--get-attr "user" entry-data) user)
+              entry-data)
+             (t
+              result))))
+   entry-set
+   :initial-value (auth-source-pass-parse-entry (car entry-set))))
+
+(defun auth-source-pass--applicable-entries (hostname user port)
+  "Return all applicable password-store entries for HOSTNAME, USER, & PORT.
+
+The result is a list of lists of password-store entries, where
+each sublist contains entries that actually exist in the
+password-store matching one of the entry name formats that
+auth-source-pass expects, most specific to least specific."
+  (let* ((possibilities-match-lists
+          (mapcar (lambda (possibility)
+                    (list (format "\\(^\\|/\\)%s$" possibility)))
+                  (auth-source-pass--generate-entry-possibilities hostname user port))))
+    (mapcar #'cdr
+            (cl-reduce (lambda (result file)
+                         (mapcar (lambda (possibility-match-list)
+                                   (if (string-match (car possibility-match-list) file)
+                                       (append possibility-match-list (list file))
+                                     possibility-match-list))
+                                 result))
+                       (auth-source-pass-entries)
+                       :initial-value possibilities-match-lists))))
+
+(defun auth-source-pass--generate-entry-possibilities (hostname user port)
+  "Return a list of possible entry path suffixes in the password-store.
+
+Based on the supported pathname patterns for HOSTNAME, USER, &
+PORT, return a list of possible suffixes for matching entries in
+the password-store."
+  (let ((base-names (auth-source-pass--name-possibilities
+                     (split-string hostname "\\."))))
+    (seq-mapcat (lambda (n)
+                  (auth-source-pass--name-port-user-possibilities n user port))
+                base-names)))
+
+(defun auth-source-pass--name-possibilities (name-components)
+  "Return a list of possible domain names matching the hostname.
+
+This function takes a list of NAME-COMPONENTS, the strings
+separated by periods in the hostname, and returns a list of full
+domain names containing the trailing sequences of thos
+components, from longest to shortest.  The shortest sequence
+returned contains two components under the assumption that no-one
+will want to associate a pass entry with an entire TLD."
+  (when (> (length name-components) 0)
```

Comment:
Good catch. It did originally function as documented but then I realized that would require changing a bunch of tests. I didn't see much harm in just checking the top-level as well even though I don't think it will be used in practice so I changed it to match all levels. I'll fix the docstring.